ash-window/src/lib.rs

```diff
@@ -9,118 +9,173 @@ use ash::{
-                .window(window.window);
-            let surface_fn = xlib_surface::Instance::new(entry, instance);
-            surface_fn.create_xlib_surface(&surface_desc, allocation_callbacks)
-        }
-
-        (RawDisplayHandle::Xcb(display), RawWindowHandle::Xcb(window)) => {
-            let surface_desc = vk::XcbSurfaceCreateInfoKHR::default()
-                .connection(
-                    display
-                        .connection
-                        .ok_or(vk::Result::ERROR_INITIALIZATION_FAILED)?
-                        .as_ptr(),
-                )
-                .window(window.window.get());
-            let surface_fn = xcb_surface::Instance::new(entry, instance);
-            surface_fn.create_xcb_surface(&surface_desc, allocation_callbacks)
-        }
-
-        (RawDisplayHandle::Android(_), RawWindowHandle::AndroidNdk(window)) => {
-            let surface_desc =
-                vk::AndroidSurfaceCreateInfoKHR::default().window(window.a_native_window.as_ptr());
-            let surface_fn = android_surface::Instance::new(entry, instance);
-            surface_fn.create_android_surface(&surface_desc, allocation_callbacks)
-        }
-
-        #[cfg(target_os = "macos")]
-        (RawDisplayHandle::AppKit(_), RawWindowHandle::AppKit(window)) => {
-            use raw_window_metal::{appkit, Layer};
-
-            let layer = match appkit::metal_layer_from_handle(window) {
-                Layer::Existing(layer) | Layer::Allocated(layer) => layer.cast(),
-            };
-
-            let surface_desc = vk::MetalSurfaceCreateInfoEXT::default().layer(&*layer);
-            let surface_fn = metal_surface::Instance::new(entry, instance);
-            surface_fn.create_metal_surface(&surface_desc, allocation_callbacks)
-        }
-
-        #[cfg(target_os = "ios")]
-        (RawDisplayHandle::UiKit(_), RawWindowHandle::UiKit(window)) => {
-            use raw_window_metal::{uikit, Layer};
+pub enum SurfaceExtension {
+    Windows(win32_surface::Instance),
+    Wayland(
+        raw_window_handle::WaylandDisplayHandle,
+        wayland_surface::Instance,
+    ),
+    Xlib(raw_window_handle::XlibDisplayHandle, xlib_surface::Instance),
+    Xcb(raw_window_handle::XcbDisplayHandle, xcb_surface::Instance),
+    Android(android_surface::Instance),
+    #[cfg(target_os = "macos")]
+    AppKit(metal_surface::Instance),
+    #[cfg(target_os = "ios")]
+    UiKit(metal_surface::Instance),
+}
 
-            let layer = match uikit::metal_layer_from_handle(window) {
-                Layer::Existing(layer) | Layer::Allocated(layer) => layer.cast(),
-            };
+impl SurfaceExtension {
+    /// Load the relevant surface extension for a given [`RawDisplayHandle`].
+    ///
+    // TODO: Safety docs about Entry/Instance lifetime?
+    ///
+    /// `instance` must have been created with platform specific surface extensions enabled, acquired
+    /// through [`enumerate_required_extensions()`].
+    pub fn load(
```

Comment:
I would prefer this name to be the same as the name used for initialization in the underlying platform specific extension instances. If they don't end up getting renamed I think it is better to use `new` here

Reply:
I have just been increasingly dissatisfied with the limited number of hours that we receive with each day, which is why the **proposed** PR for renaming them all to `load()` wasn't opened yet. That change is strictly for another pull request.

---

> If we rename this away from `...Extension` as I propose elsewhere, `load` might be less clear even after the function pointer table constructors are renamed.

Would you recommend keeping it `fn new()` even after I rename the loader constructors... to `fn load()`?

Reply:
I think `load` makes sense if design the API to visibly be "some Vulkan extension", and `new` makes sense if we design the API to be a black box. I favor the latter for better encapsulation.